networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/topics/validation/VoluntaryExitValidatorTest.java

```diff
@@ -0,0 +1,175 @@
+  private VoluntaryExitValidator voluntaryExitValidator;
+
+  @BeforeAll
+  static void beforeAll() {
+    Constants.SLOTS_PER_EPOCH = 2;
+    Constants.EPOCHS_PER_ETH1_VOTING_PERIOD = 1;
+    Constants.PERSISTENT_COMMITTEE_PERIOD = 2;
+  }
+
+  @AfterAll
+  static void afterAll() {
+    Constants.setConstants("minimal");
+  }
+
+  @BeforeEach
+  void beforeEach() {
+    recentChainData = MemoryOnlyRecentChainData.create(new EventBus());
+    beaconChainUtil = BeaconChainUtil.create(recentChainData, VALIDATOR_KEYS, true);
+    voluntaryExitValidator = new VoluntaryExitValidator(recentChainData);
+  }
+
+  @Test
+  public void shouldReturnValidForValidVoluntaryExit() throws Exception {
+    beaconChainUtil.initializeStorage();
+    beaconChainUtil.createAndImportBlockAtSlot(6);
+    SignedVoluntaryExit exit =
+        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 3);
+    assertThat(voluntaryExitValidator.validate(exit)).isEqualTo(VALID);
+  }
+
+  @Test
+  public void shouldReturnInvalidForExitsAfterTheFirstForValidator() throws Exception {
+    beaconChainUtil.initializeStorage();
+    beaconChainUtil.createAndImportBlockAtSlot(6);
+    SignedVoluntaryExit exit1 =
+        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 3);
+
+    SignedVoluntaryExit exit2 =
+        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 3);
+
+    SignedVoluntaryExit exit3 =
+        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 3);
+
+    assertThat(voluntaryExitValidator.validate(exit1)).isEqualTo(VALID);
+    assertThat(voluntaryExitValidator.validate(exit2)).isEqualTo(INVALID);
+    assertThat(voluntaryExitValidator.validate(exit3)).isEqualTo(INVALID);
+  }
+
+  @Test
+  public void shouldReturnInvalidForExitWithInvalidSignature() throws Exception {
+    beaconChainUtil.initializeStorage();
+    beaconChainUtil.createAndImportBlockAtSlot(6);
+    SignedVoluntaryExit exit1 =
+        voluntaryExitGenerator.withInvalidSignature(
+            recentChainData.getBestState().orElseThrow(), 3);
+    assertThat(voluntaryExitValidator.validate(exit1)).isEqualTo(INVALID);
+  }
+
+  @Test
+  public void shouldReturnInvalidForExitOfInactiveValidator() throws Exception {
+    TestDepositGenerator testDepositGenerator =
+        new TestDepositGenerator(VALIDATOR_KEYS.subList(0, 20));
+    List<Deposit> deposits = testDepositGenerator.getDeposits(10, 20, 20);
+    initializeStorage(recentChainData, VALIDATOR_KEYS.subList(0, 10));
+
+    beaconChainUtil.setEth1DataOfChain(
+        new Eth1Data(testDepositGenerator.getRoot(), UnsignedLong.valueOf(20), Bytes32.ZERO));
+    beaconChainUtil.createAndImportBlockAtSlotWithDeposits(UnsignedLong.valueOf(6), deposits);
+
+    SignedVoluntaryExit exit =
+        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 15);
+    assertThat(voluntaryExitValidator.validate(exit)).isEqualTo(INVALID);
+  }
```

Comment:
You can do this without messing around with additional deposits and Eth1Data by just adding a validator to the genesis state that hasn't deposited enough to be activated:
```suggestion
  @Test
  public void shouldReturnInvalidForExitOfInactiveValidator() throws Exception {
    final DepositGenerator depositGenerator = new DepositGenerator(true);
    final List<DepositData> initialDepositData =
        new MockStartDepositGenerator(depositGenerator)
            .createDeposits(VALIDATOR_KEYS.subList(0, 10));
    // Add an inactive validator (they haven't deposited enough to become a validator)
    final BLSKeyPair inactiveValidatorKeyPair = VALIDATOR_KEYS.get(10);
    initialDepositData.add(depositGenerator.createDepositData(inactiveValidatorKeyPair, UnsignedLong.ONE, inactiveValidatorKeyPair.getPublicKey()));
    final BeaconState genesisState = new MockStartBeaconStateGenerator()
        .createInitialBeaconState(UnsignedLong.ZERO, initialDepositData);
    recentChainData.initializeFromGenesis(genesisState);

    SignedVoluntaryExit exit =
        voluntaryExitGenerator.valid(recentChainData.getBestState().orElseThrow(), 10, false);
    assertThat(voluntaryExitValidator.validate(exit)).isEqualTo(INVALID);
  }
```

Reply:
Do you think I should keep `setEth1DataOfChain` as a utility method, or remove it since this PR won't require it anymore?

Reply:
Probably take it out of this PR but keep it around locally - you're going to want it when fixing the bug about using the wrong Eth1Data when selecting deposits to go in the block. :)

Reply:
Yep, this approach is way smarter and simpler btw.

Reply:
Done.